web/src/message_view.js

```diff
@@ -661,6 +661,86 @@ export function show(raw_terms, opts) {
     }
 }
 
+function handle_anchor_case(anchor, has_found_anchor, get_msg_id) {
+    // These functions are scoped inside `handle_anchor_case` to
+    // to avoid them being used for any other purpose.
+    function duplicate_current_msg_list_with_new_data(data) {
+        const msg_list = new message_list.MessageList({data});
+        msg_list.reading_prevented = message_lists.current.reading_prevented;
+        return msg_list;
+    }
+
+    function select_msg_id(msg_id, select_opts) {
+        message_lists.current.select_id(msg_id, {
+            then_scroll: true,
+            from_scroll: false,
+            ...select_opts,
+        });
+    }
+
+    function select_anchor_using_data(data) {
+        const msg_list = duplicate_current_msg_list_with_new_data(data);
+        message_lists.update_current_message_list(msg_list);
+        // `force_rerender` is required to render the new data.
+        select_msg_id(get_msg_id(), {force_rerender: true});
+    }
+
+    if (has_found_anchor(message_lists.current.data.fetch_status)) {
+        select_msg_id(get_msg_id());
+    } else {
+        let msg_list_data = new MessageListData({
+            filter: message_lists.current.data.filter,
+            excludes_muted_topics: message_lists.current.data.excludes_muted_topics,
+        });
+        load_local_messages(msg_list_data);
+
+        if (has_found_anchor(all_messages_data.fetch_status)) {
```

Comment:
Why is this `all_messages_data`, rather than the `msg_list_data` that we just constructed?

I guess for most users, we expect `has_found_anchor(all_messages_data.fetch_status)` to be always false for the `"oldest"` check in old enough accounts, but maybe that's OK because if we have the data, the oldest `has_found_anchor` check will work, and

Reply:
Sorry, I should have added a comment here. Modified the code here a bit and added this comment:

```
        // We can load messages into `msg_list_data` but we don't know
        // the fetch status until we contact server. If we are contacting the
        // server, it is better to just fetch the required messages instead
        // of just fetching status.
        //
        // So, a cheaper check is to see if we have found the anchor in
        // `all_messages_data`, and if we have, we can say `msg_list_data`
        // will also have the anchor (for oldest / newest anchors at least).
````